reconcile/utils/oc_map.py

```diff
@@ -0,0 +1,317 @@
+import logging
+from threading import Lock
+from typing import (
+    Any,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Protocol,
+    Sequence,
+    Union,
+)
+
+from sretoolbox.utils import threaded
+
+from reconcile.utils.oc import (
+    OC,
+    JumpHostSSH,
+    OCDeprecated,
+    OCLogMsg,
+    StatusCodeError,
+)
+from reconcile.utils.secret_reader import (
+    HasSecret,
+    SecretNotFound,
+    SecretReaderBase,
+)
+
+
+class HasDisable(Protocol):
+    integrations: Optional[list[str]]
+    e2e_tests: Optional[list[str]]
+
+
+class HasCluster(Protocol):
+    name: str
+    server_url: str
+    internal: Optional[bool]
+    insecure_skip_tls_verify: Optional[bool]
+
+    @property
+    def automation_token(self) -> Optional[HasSecret]:
+        ...
+
+    @property
+    def cluster_admin_automation_token(self) -> Optional[HasSecret]:
+        ...
+
+    @property
+    def disable(self) -> Optional[HasDisable]:
+        ...
+
+
+class HasNamespace(Protocol):
+    cluster: HasCluster
+    cluster_admin: Optional[bool]
+
+
+class OCMap:
+    """OCMap gets a GraphQL query results list as input
+    and initiates a dictionary of OC clients per cluster.
+
+    The input must contain either 'clusters' or 'namespaces', but not both.
+
+    In case a cluster does not have an automation token
+    the OC client will be initiated with a OCLogMessage.
+    """
+
+    def __init__(
+        self,
+        clusters: Optional[Sequence[HasCluster]] = None,
+        clusters_untyped: Optional[Sequence[MutableMapping[Any, Any]]] = None,
+        namespaces: Optional[Sequence[HasNamespace]] = None,
+        namespaces_untyped: Optional[Sequence[MutableMapping[Any, Any]]] = None,
```

Comment:
providing both typed and untyped data here irritates me a bit.

have you considered to drop support of untyped clusters and namespaces in this copy `OC_Map` instead of offering the mixed mode? since the old version still lives in `oc.py` existing usage is unaffected and integrations wanting to adopt the new version need to be touched anyways and can adhere to the typed approach during that process.

Reply:
The untyped data is currently needed for the underlying classes `OCDeprecated`, `OCNative` and `SSHJumphost`. The idea was to not refactor them all in a single PR, but rather split this up into multiple (working) PRs.

Basically I wanted to create follow-up PRs shortly after this one in which I will refactor those underlying classes. But for this first step to be functional, I still need to pass the old nested dictionaries to work with the old classes.

Reply:
ah makes sense. thanks for the details!